File: src/inspect_models.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/dnn.hpp>
#include <iostream>

using namespace cv;
using namespace cv::dnn;
using namespace std;

void inspect(string path) {
    cout << "Inspecting: " << path << endl;
    Net net = readNetFromONNX(path);
    if (net.empty()) {
        cout << "Failed to load net" << endl;
        return;
    }
    
    vector<String> outLayerNames = net.getUnconnectedOutLayersNames();
    cout << "Outputs:" << endl;
    for (const auto& name : outLayerNames) {
        cout << "  - " << name << endl;
    }
    
    // Check first layer (usually input)
    vector<String> layerNames = net.getLayerNames();
    if (!layerNames.empty()) {
        cout << "First 5 layers:" << endl;
        for (int i = 0; i < min((int)layerNames.size(), 5); ++i) {
            cout << "  - " << layerNames[i] << endl;
        }
    }
}

int main(int argc, char** argv) {
    if (argc > 1) {
        inspect(argv[1]);
    } else {
        inspect("../models/palm_detection.onnx");
        inspect("../models/handpose_estimation.onnx");
    }
    return 0;
}

File: src/inspect_models.py
import cv2
import numpy as np

def inspect(path, input_size):
    print(f"Inspecting: {path}")
    try:
        net = cv2.dnn.readNetFromONNX(path)
        blob = cv2.dnn.blobFromImage(np.zeros((input_size, input_size, 3), np.uint8), 1.0/255.0, (input_size, input_size))
        
        # Transpose NCHW -> NHWC
        blob_nhwc = np.transpose(blob, (0, 2, 3, 1))
        
        net.setInput(blob_nhwc)
        out_names = net.getUnconnectedOutLayersNames()
        outs = net.forward(out_names)
        
        for i, out in enumerate(outs):
            print(f"  Output {i} ({out_names[i]}): {out.shape}")
            # Print a few values to see range
            print(f"    Sample: {out.flatten()[:20]}")
            
    except Exception as e:
        print(f"Error: {e}")
    print("-" * 20)

inspect("/home/zius/Projects/KineticFPS/models/palm_detection.onnx", 128)
inspect("/home/zius/Projects/KineticFPS/models/handpose_estimation.onnx", 224)
